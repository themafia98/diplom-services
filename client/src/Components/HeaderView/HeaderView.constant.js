const MARGIN_TAB = 5;

export { MARGIN_TAB };
