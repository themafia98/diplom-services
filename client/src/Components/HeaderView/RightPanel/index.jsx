import React from "react";
import Updater from "../../Updater";
import UserPopup from "../../UserPopup";

import Scrollbars from "react-custom-scrollbars";
import NotificationPopup from "../../NotificationPopup";
import Status from "./Status";

const RightPanel = ({ onUpdate, onLogout, goCabinet, status, shouldUpdate }) => {
    return (
        <div className="headerControllers rightPanel">
            <UserPopup goCabinet={goCabinet} />
            <div className="groupControllers">
                <NotificationPopup />
                <Updater onClick={onUpdate} additionalClassName="updaterDefault" />
                <div onClick={onLogout} className="logout">
                    Выйти
                </div>
                <Status shouldUpdate={shouldUpdate} statusApp={status} />
            </div>
        </div>
    );
};
export default RightPanel;
