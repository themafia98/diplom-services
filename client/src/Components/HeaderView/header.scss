header {
    $boxTab: 80%;

    &.ant-layout-header {
        padding: 0;
        background-color: $headerColor;
    }
    display: flex;
    flex-direction: row;
    align-items: center;

    .wrapperTabs {
        width: $boxTab;
    }

    ul {
        &.tabsMenu {
            display: flex;
            align-items: center;
            width: $boxTab;
            height: 100%;
            padding: 0;
            margin: 0;
            li {
                margin: 0 5px;
                cursor: pointer;
                span {
                    cursor: pointer;
                }
            }

            .closeTab {
                position: absolute;
                top: 3%;
                left: 95%;
                transform: translate(-50%);
                color: $headerColor;
                font-size: 15px;
                &.resizeTab {
                    left: 90%;
                }

                &:hover {
                    color: red;
                }
            }
            li {
                border: 1px solid $tabBorder;
                position: relative;
                &.active {
                    background-color: $tab;
                }

                display: flex;
                justify-content: center;
                align-items: center;
                height: 65%;
                cursor: pointer;
                background-color: $unactiveTab;
                padding: 0;
                // flex: 0 0 10%;
                list-style-type: none;
                margin: 0 5px;

                span {
                    font-weight: bold;
                }

                .tabWrapper-content {
                    height: 100%;
                    overflow: hidden;
                    width: 90%;
                    align-items: center;
                    justify-content: center;
                    display: flex;
                    white-space: nowrap;
                    .tab-content {
                        font-weight: bold;
                        text-align: center;
                        overflow: hidden;
                        display: block;
                        white-space: nowrap;
                        width: 90%;

                        user-select: none;
                        text-overflow: ellipsis;
                    }
                    &.selected {
                        color: $selected;
                    }
                }
            }
        }
    }
}

.statusApp {
    position: absolute;
    top: 0;
    right: 0;
    display: flex;
    .statusAppIcon {
        width: 15px;
        height: 15px;
        margin-left: 5px;
        border-radius: 50%;
        border: 1px solid black;
        top: 5px;
        right: 5px;
        &.online {
            background-color: green;
        }
        &.offline {
            background-color: red;
        }
    }
}

@media screen and (max-width: 1351px) {
    header {
        .headerControllers {
            &.rightPanel {
                width: 35%;
            }
        }
    }
}

@media screen and (max-width: 914px) {
    header {
        .headerControllers {
            &.rightPanel {
                width: 45%;
            }
        }
    }
}

@media screen and (max-width: 820px) {
    header {
        .headerControllers {
            &.rightPanel {
                width: 60%;
            }
        }
    }
}

@media screen and (max-width: 680px) {
    header {
        .headerControllers {
            &.rightPanel {
                .userPopup {
                    display: none;
                }
                width: 35%;
            }
        }
    }
}

@media screen and (max-width: 420px) {
    header {
        .headerControllers {
            &.rightPanel {
                .notificationControllers {
                    display: none;
                }
            }
        }
    }
}
