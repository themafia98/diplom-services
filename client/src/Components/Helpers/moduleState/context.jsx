import { createContext } from 'react';

const ModuleContextState = createContext();
const ModuleContextActions = createContext();

export { ModuleContextState, ModuleContextActions };
