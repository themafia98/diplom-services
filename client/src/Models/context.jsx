import { createContext } from 'react';
import config from '../config.json';
import Request from '../Models/Rest';
import Schema from '../Models/Schema';
import Notification from '../Models/NotificationRouter';
import { clientDB } from './ClientSideDatabase';

export const modelMethods = {
  clientDB,
  rest: new Request(),
  Request,
  config,
  schemaModel: Schema,
  schema: new Schema('no-strict'),
  Notification,
};

const modelsContext = createContext(modelMethods);

export default modelsContext;
