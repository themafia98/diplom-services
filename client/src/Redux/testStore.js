import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

/** stores for  jest */

export const initialState = {
  tabReducer: {
    parentSize: null,
    childrenSize: null,
    flag: false,
  },
  publicReducer: {
    status: 'online',
    prewStatus: 'online',
    requestError: null,
    caches: {},
  },
  router: {
    currentActionTab: 'mainModule',
    actionTabs: ['MainModule'],
    routeDataActive: {},
    routeData: {},
  },
};

const middlewares = [thunk.withExtraArgument({})]; /** test middlewares */
const mockStore = configureStore(middlewares); /** test config store */

const store = mockStore(initialState); /** init test store */

export default store;
