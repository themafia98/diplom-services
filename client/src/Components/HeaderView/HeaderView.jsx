import React, { useState, useRef, memo, useEffect } from 'react';
import { DragDropContext, Droppable } from 'react-beautiful-dnd';
import { useDispatch, useSelector } from 'react-redux';
import { Layout } from 'antd';
import { dragEndTabAction } from 'Redux/actions/tabActions';
import RightPanel from './RightPanel';
import Tab from './Tab/index';
import { MARGIN_TAB } from './HeaderView.constant';
import { headerViewType } from './HeaderView.types';

const { Header } = Layout;

const HeaderView = memo(
  ({ tabs, goCabinet, logout, activeTabEUID, cbMenuTabHandler, dashboardStream, webSocket }) => {
    const dispatch = useDispatch();

    const [length, setLength] = useState(1);
    const [size, setSize] = useState(160);
    const [sizeParent, setSizeParent] = useState(null);

    const tabsMenuRef = useRef(null);
    const { current: tabsMenuNode = null } = tabsMenuRef || {};

    const { shouldUpdate, status, udata, appConfig } = useSelector((state) => {
      const { status, udata, appConfig } = state.publicReducer;
      const { shouldUpdate = false } = state.router;
      return {
        shouldUpdate,
        status,
        udata,
        appConfig,
      };
    });

    useEffect(() => {
      const sizes = sizeParent / tabs.length - MARGIN_TAB;
      const counter = ~~(sizeParent / size);

      if (sizeParent === null && tabsMenuNode) {
        setLength(tabs.length);
        setSizeParent(tabsMenuNode.getBoundingClientRect().width);
      }

      if ((tabs.length >= counter && sizeParent !== null) || (tabs.length < length && size < 160)) {
        setLength(tabs.length);
        setSize(sizes);
      }
    }, [sizeParent, size, tabsMenuNode, length, tabs]);

    const reorder = (list, dragIndex, dropIndex) => {
      const [removed] = list.splice(dragIndex, 1);
      list.splice(dropIndex, 0, removed);
      return list;
    };

    const onDragEnd = ({ destination, source }, tabsList) => {
      const { index: indexDest } = destination || {};
      const { index } = source || {};

      if (!destination) return;

      dispatch(dragEndTabAction(reorder([...tabsList], index, indexDest)));
    };

    const renderTabs = (items) => {
      return (
        <DragDropContext onDragEnd={(event) => onDragEnd(event, items)}>
          <Droppable direction="horizontal" droppableId="droppable">
            {(provided, snapshot) => (
              <div className="droppable-wrapper" ref={provided.innerRef}>
                <ul ref={tabsMenuRef} {...provided.droppableProps} className="tabsMenu">
                  {items.map(({ EUID = '', VALUE = '' }, index) => {
                    return (
                      <Tab
                        hendlerTab={cbMenuTabHandler}
                        active={activeTabEUID === EUID}
                        key={EUID}
                        itemKey={EUID}
                        value={VALUE}
                        sizeTab={size}
                        index={index}
                      />
                    );
                  })}
                </ul>
              </div>
            )}
          </Droppable>
        </DragDropContext>
      );
    };

    const update = () => {
      if (dashboardStream) dashboardStream.emit('EventUpdate', true);
    };

    const notificationDep = {
      streamStore: 'streamList',
      streamModule: 'system',
      store: 'redux',
      filterStream: 'uidCreater',
    };

    return (
      <Header>
        {tabs ? renderTabs(tabs) : null}
        <RightPanel
          appConfig={appConfig}
          udata={udata}
          shouldUpdate={shouldUpdate}
          status={status}
          goCabinet={goCabinet}
          onLogout={logout}
          onUpdate={update}
          notificationDep={notificationDep}
          webSocket={webSocket}
        />
      </Header>
    );
  },
);

HeaderView.defaultProps = {
  goCabinet: null,
};

HeaderView.propTypes = headerViewType;

export default HeaderView;
