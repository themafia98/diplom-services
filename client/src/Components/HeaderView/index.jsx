import React from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import _ from "lodash";
import { Layout } from "antd";

import { setParentSizeAction, setChildrenSizeAction } from "../../Redux/actions/tabActions";
import Tab from "./Tab";
import RightPanel from "./RightPanel";

const { Header } = Layout;

class HeaderView extends React.PureComponent {
    state = {
        defaultSizeTab: 160
    };

    static propTypes = {
        dashboardStrem: PropTypes.object.isRequired,
        cbMenuTabHandler: PropTypes.func.isRequired,
        activeTabEUID: PropTypes.string.isRequired,
        actionTabs: PropTypes.oneOfType([PropTypes.array, PropTypes.bool]).isRequired,
        logout: PropTypes.func.isRequired
    };

    componentDidUpdate = () => {
        const { tabArray = [], tabReducer = {}, onSetChildrenSizeAction, onsetParentSizeAction } = this.props;
        let sizeTab = !tabReducer.flag ? this.state.defaultSizeTab : tabReducer.childrenSize;
        if (!_.isNull(this.wrapper) && !_.isNull(tabReducer) && _.isNull(tabReducer.parentSize)) {
            const newSize = this.wrapper.getBoundingClientRect().width;
            if (tabReducer.parentSize !== newSize) onsetParentSizeAction(newSize);
        }

        if (
            !_.isNull(tabReducer) &&
            !_.isNull(tabReducer.childrenSize) &&
            !_.isNull(tabReducer.parentSize) &&
            tabArray.length > 1
        ) {
            const length = tabReducer.childrenSize * tabArray.length;
            if (tabReducer.parentSize < length + sizeTab) {
                const size = tabReducer.parentSize / tabArray.length;
                if (size !== tabReducer.childrenSize) onSetChildrenSizeAction(size, true);
            }
        }
    };

    wrapper = null;
    refWrapper = node => (this.wrapper = node);

    renderTabs = items => {
        const { activeTabEUID = "mainModule", cbMenuTabHandler, tabArray = [], tabReducer = {} } = this.props;
        let sizeTab = !tabReducer.flag ? this.state.defaultSizeTab : tabReducer.childrenSize;
        let flag = false;
        const length = sizeTab * tabArray.length;
        if (sizeTab > tabReducer.childrenSize && tabReducer.parentSize < length + sizeTab) {
            flag = true;
        }

        return (
            <ul ref={this.refWrapper} className="tabsMenu">
                {items.map(item => {
                    return (
                        <Tab
                            hendlerTab={cbMenuTabHandler}
                            active={activeTabEUID === item.EUID}
                            key={item.EUID}
                            itemKey={item.EUID}
                            value={item.VALUE}
                            sizeTab={sizeTab}
                            flag={flag}
                        />
                    );
                })}
            </ul>
        );
    };

    update = () => {
        const { dashboardStrem } = this.props;
        dashboardStrem.emit("EventUpdate", true);
    };

    render() {
        const { actionTabs = ["mainModule"], goCabinet, status, shouldUpdate } = this.props;

        return (
            <Header>
                {actionTabs ? this.renderTabs(actionTabs) : null}
                <RightPanel
                    shouldUpdate={shouldUpdate}
                    status={status}
                    goCabinet={goCabinet}
                    onLogout={this.logout}
                    onUpdate={this.update}
                />
            </Header>
        );
    }
}

const mapStateTopProps = state => {
    const { status = "online" } = state.publicReducer;
    const { shouldUpdate = false } = state.router;
    return {
        tabReducer: state.tabReducer,
        tabArray: state.router.actionTabs,
        shouldUpdate,
        status
    };
};

const mapDispatchToProps = dispatch => {
    return {
        onsetParentSizeAction: size => dispatch(setParentSizeAction(size)),
        onSetChildrenSizeAction: (size, flag) => dispatch(setChildrenSizeAction(size, flag))
    };
};

export default connect(mapStateTopProps, mapDispatchToProps)(HeaderView);
export { HeaderView };
