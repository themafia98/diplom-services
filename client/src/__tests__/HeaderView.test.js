import React from 'react';
import { EventEmitter } from 'events';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import { HeaderView } from '../Components/HeaderView/HeaderView';
import { initialState } from '../Redux/testStore';
import context from '../Models/context';
describe('<HeaderView /> and children', () => {
  test('Should render', () => {
    const props = {
      ...initialState,
      dashboardStrem: new EventEmitter(),
      cbMenuTabHandler: () => {},
      activeTabs: [{ EUID: 'mainModule' }],
      activeTabEUID: 'mainModule',
      logout: () => {},
    };

    const HeaderViewWrapper = shallow(<HeaderView {...props} />, { context });
    expect(toJson(HeaderViewWrapper)).toMatchSnapshot();

    expect(HeaderViewWrapper.find('.tabsMenu').exists()).toBeTruthy();

    const nextProps = {
      ...props,
      activeTabs: [{ EUID: 'mainModule1' }],
    };

    const HeaderViewWrapperNewProps = shallow(<HeaderView {...nextProps} />);
    expect(HeaderViewWrapperNewProps).toMatchSnapshot();
  });
});
