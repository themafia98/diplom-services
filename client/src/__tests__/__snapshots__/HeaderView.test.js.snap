// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HeaderView /> and children Should render 1`] = `
<Header>
  <ul
    className="tabsMenu"
  >
    <Tab
      active={true}
      hendlerTab={[Function]}
      itemKey="mainModule"
      key="mainModule"
      sizeTab={160}
    />
  </ul>
  <RightPanel
    active={true}
    goCabinet={null}
    notificationDep={
      Object {
        "filterStream": "uidCreater",
        "onSaveComponentState": undefined,
        "store": "redux",
        "streamModule": "system",
        "streamStore": "streamList",
      }
    }
    onLogout={[Function]}
    onUpdate={[Function]}
    shouldUpdate={false}
    status=""
    udata={Object {}}
  />
</Header>
`;

exports[`<HeaderView /> and children Should render 2`] = `
<Header>
  <ul
    className="tabsMenu"
  >
    <Tab
      active={false}
      hendlerTab={[Function]}
      itemKey="mainModule1"
      key="mainModule1"
      sizeTab={160}
    />
  </ul>
  <RightPanel
    active={true}
    goCabinet={null}
    notificationDep={
      Object {
        "filterStream": "uidCreater",
        "onSaveComponentState": undefined,
        "store": "redux",
        "streamModule": "system",
        "streamStore": "streamList",
      }
    }
    onLogout={[Function]}
    onUpdate={[Function]}
    shouldUpdate={false}
    status=""
    udata={Object {}}
  />
</Header>
`;
